pkg/resource/deploy/step.go

```diff
@@ -709,6 +709,135 @@ func (s *RefreshStep) Apply(preview bool) (resource.Status, StepCompleteFunc, er
+	replacing    bool                           // true if we are replacing a Pulumi-managed resource.
+	diffs        []resource.PropertyKey         // any keys that differed between the user's program and the actual state.
+	detailedDiff map[string]plugin.PropertyDiff // the structured property diff.
+}
+
+func NewImportStep(plan *Plan, reg RegisterResourceEvent, new *resource.State) Step {
+	contract.Assert(new != nil)
+	contract.Assert(new.URN != "")
+	contract.Assert(new.ID != "")
+	contract.Assert(new.Custom)
+	contract.Assert(!new.Delete)
+	contract.Assert(!new.External)
+
+	return &ImportStep{
+		plan: plan,
+		reg:  reg,
+		new:  new,
+	}
+}
+
+func NewImportReplacementStep(plan *Plan, reg RegisterResourceEvent, original, new *resource.State) Step {
+	contract.Assert(original != nil)
+	contract.Assert(new != nil)
+	contract.Assert(new.URN != "")
+	contract.Assert(new.ID != "")
+	contract.Assert(new.Custom)
+	contract.Assert(!new.Delete)
+	contract.Assert(!new.External)
+
+	return &ImportStep{
+		plan:      plan,
+		reg:       reg,
+		original:  original,
+		new:       new,
+		replacing: true,
+	}
+}
+
+func (s *ImportStep) Op() StepOp {
+	if s.replacing {
+		return OpImportReplacement
+	}
+	return OpImport
+}
+
+func (s *ImportStep) Plan() *Plan                                  { return s.plan }
+func (s *ImportStep) Type() tokens.Type                            { return s.new.Type }
+func (s *ImportStep) Provider() string                             { return s.new.Provider }
+func (s *ImportStep) URN() resource.URN                            { return s.new.URN }
+func (s *ImportStep) Old() *resource.State                         { return s.old }
+func (s *ImportStep) New() *resource.State                         { return s.new }
+func (s *ImportStep) Res() *resource.State                         { return s.new }
+func (s *ImportStep) Logical() bool                                { return !s.replacing }
+func (s *ImportStep) Diffs() []resource.PropertyKey                { return s.diffs }
+func (s *ImportStep) DetailedDiff() map[string]plugin.PropertyDiff { return s.detailedDiff }
+
+func (s *ImportStep) Apply(preview bool) (resource.Status, StepCompleteFunc, error) {
+	complete := func() { s.reg.Done(&RegisterResult{State: s.new}) }
+
+	// Read the current state of the resource to import. If the provider does not hand us back any inputs for the
+	// resource, it probably needs to be updated. If the resource does not exist at all, fail the import.
+	prov, err := getProvider(s)
+	if err != nil {
+		return resource.StatusOK, nil, err
+	}
+	read, rst, err := prov.Read(s.new.URN, s.new.ID, nil, nil)
+	if err != nil {
+		if initErr, isInitErr := err.(*plugin.InitError); isInitErr {
+			s.new.InitErrors = initErr.Reasons
+		} else {
+			return rst, nil, err
+		}
+	}
+	if read.Outputs == nil {
+		return rst, nil, errors.Errorf("resource '%v' does not exist", s.new.ID)
+	}
+	if read.Inputs == nil {
+		return resource.StatusOK, nil, errors.Errorf(
```

Comment:
Just wondering - do our existing released plugins today fail in this way, or is it just much older versions?

Reply:
Only older versions will fail in this fashion. FWIW, I considered adding something to `ConfigureResponse` to indicate that a provider does support importing resources, which would allow a provider to return a nil input bag here if it can properly support `Check` and `Diff` without old inputs.